pageserver/src/tenant/mgr.rs

```diff
@@ -864,34 +876,185 @@ pub(crate) enum GetTenantError {
 ///
 /// This method is cancel-safe.
-pub(crate) async fn get_tenant(
+pub(crate) fn get_tenant(
     tenant_id: TenantId,
     active_only: bool,
 ) -> Result<Arc<Tenant>, GetTenantError> {
-    let m = TENANTS.read().await;
-    let tenant = m
-        .get(&tenant_id)
-        .ok_or(GetTenantError::NotFound(tenant_id))?;
-
-    match tenant.current_state() {
-        TenantState::Broken {
-            reason,
-            backtrace: _,
-        } if active_only => Err(GetTenantError::Broken(reason)),
-        TenantState::Active => Ok(Arc::clone(tenant)),
-        _ => {
-            if active_only {
-                Err(GetTenantError::NotActive(tenant_id))
+    let locked = TENANTS.read().unwrap();
+    let peek_slot = tenant_map_peek_slot(&locked, &tenant_id, true)?;
+
+    match peek_slot {
+        Some(TenantSlot::Attached(tenant)) => match tenant.current_state() {
+            TenantState::Broken {
+                reason,
+                backtrace: _,
+            } if active_only => Err(GetTenantError::Broken(reason)),
+            TenantState::Active => Ok(Arc::clone(tenant)),
+            _ => {
+                if active_only {
+                    Err(GetTenantError::NotActive(tenant_id))
+                } else {
+                    Ok(Arc::clone(tenant))
+                }
+            }
+        },
+        Some(TenantSlot::InProgress(_)) => Err(GetTenantError::NotActive(tenant_id)),
+        None | Some(TenantSlot::Secondary) => Err(GetTenantError::NotFound(tenant_id)),
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub(crate) enum GetActiveTenantError {
+    /// We may time out either while TenantSlot is InProgress, or while the Tenant
+    /// is in a non-Active state
+    #[error(
+        "Timed out waiting {wait_time:?} for tenant active state. Latest state: {latest_state:?}"
+    )]
+    WaitForActiveTimeout {
+        latest_state: Option<TenantState>,
+        wait_time: Duration,
+    },
+
+    /// The TenantSlot is absent, or in secondary mode
+    #[error(transparent)]
+    NotFound(#[from] GetTenantError),
+
+    /// Cancellation token fired while we were waiting
+    #[error("cancelled")]
+    Cancelled,
+
+    /// Tenant exists, but is in a state that cannot become active (e.g. Stopping, Broken)
+    #[error("will not become active.  Current state: {0}")]
+    WillNotBecomeActive(TenantState),
+}
+
+enum TimeoutCancellableError {
+    Timeout,
+    Cancelled,
+}
+
+/// Wrap [`tokio::time::timeout`] with a CancellationToken.
+async fn timeout_cancellable<F>(
+    duration: Duration,
+    future: F,
+    cancel: &CancellationToken,
+) -> Result<F::Output, TimeoutCancellableError>
+where
+    F: std::future::Future,
+{
+    tokio::select!(
+        r = tokio::time::timeout(duration, future) => {
+            r.map_err(|_| TimeoutCancellableError::Timeout)
+
+        },
+        _ = cancel.cancelled() => {
+            Err(TimeoutCancellableError::Cancelled)
+
+        }
+    )
+}
```

Comment:
This feels like it should be in `utils` 

Regarding the signature, I think the order `duration, cancel, future` would be better for other use cases where we pass `future = async move { ... }` inline.

Reply:
I'm ambivalent about the arg order: this way is more aligned with what tokio::timeout looks like, but you have a fair point about async{} users.  Changed it.

Generally I don't put things in utils until they're really used more than one place, but this one feels like it inevitably will be, so sure.

ce82103a7